### ISSUE TITLE
onErrorReturnWith on a Firestore snapshots() requires reinstalling app for further Firestore updates

### ISSUE THREAD

**Issue by misteroak** [ReactiveX/rxdart#616]
Hi - I'm far from an expert in using rxdart and Firestore, so appoligze in advance if I'm missing something basic here. I'm experiencing a weird behavior with onErrorReturnWith in the following code (I'm following [Reso Coder's DDD and Firebase tutorial series](https://github.com/ResoCoder/flutter-ddd-firebase-course)):

```
Stream<Either<NoteFailure, KtList<Note>>> watchAll() async* {
    final userDoc = await _firestore.userDocument();

    yield* userDoc.noteCollection.snapshots().map(
      (snapshot) {
        return right<NoteFailure, KtList<Note>>(
          snapshot.docs.map((doc) => NoteDto.fromFirestore(doc).toDomain()).toImmutableList(),
        );
      },
    ).onErrorReturnWith((exception, t) {
      if (exception is PlatformException && exception.message!.contains('PERMISSION_DENIED')) {
        return left(const NoteFailure.insufficientPermissions());
      } else {
        //TODO: Log error - log.error(e.toString())
        return left(const NoteFailure.unexpected());
      }
    });
  }
```

I noticed that if there is an error inside the map function (say an invalid json or something), onErrorReturnWith() is called and is returning a failure as expected. However, from that point on, the Firestore data is no longer updating on the client.  A hot restart and even after killing the app on the phone and debugging again won't fix. Only after removing the app from the phone it starts working again.

So the steps I'm seeing:
1. Load data from Firestore (✅ works as expected)
2. Make changes to data on Firestore Console, with no errors (✅ data updates successfully on client)
3. Make further changes to data, but this with an error, e.g., missing required field which will cause the .toDomain() call to break. (✅  data on clients updates successfully and the error is caught)
4. Fix the problem on Firestore so it's the same data as in step # 2 (❌ - data not updated on client)
5. Try hot restart (❌ - data not updated on client)
6. Try closing the app on the phone/simulator and start debugging again (❌ - data not updated on client)
7. Delete the app on the phone/simulator and start debugging again (✅ data updates successfully on client)

Now, if I remove the onErrorReturnWith call and instead I use ..onError() on the stream listener that subscribes to this stream, everything works as expected - I can "break" stuff on the Firebase Console and fix them and everything updates on the client as expected.

**Comment by narcodico** [comment]
Hi @misteroak 👋

I wouldn't use `onErrorReturnWith` because like you're facing, once an error occurs, the subscription will be done.

You can always use `handleError` and deal with it in a safe fashion or write your own `StreamTransformer` which adds a valid value to the sink whenever an error occurs. Up to you and your use case.

**Comment by hoc081098** [comment]
I think that is unexpected behavior, I will fix it soon 👍

**Comment by misteroak** [comment]
Thank you @hoc081098!
